### ISSUE TITLE
Large files cause panic with S3

### ISSUE THREAD

**Issue by mousedownmike** [aws/aws-xray-sdk-go#175]
I have a simple service that pulls content from a URL and uploads it to S3.  For smallish files this works fine but for larger files, I consistently get:

`panic: failed to begin subsegment named 's3': segment cannot be found.`

I have created a [sample function that illustrates the problem](https://github.com/mousedownmike/aws-xray-s3-go-panic).  The README has the full panic output as well.

https://github.com/mousedownmike/aws-xray-s3-go-panic

**Comment by mousedownmike** [comment]
I think this might be related to #152 .  

Should this failure make that issue a bug instead of a feature request?  Basically the developer has no idea why calling the same API would succeed sometimes yet fail other times since the call to SignRequest is happening in the AWS Go SDK and not in the developer's code.

**Comment by bhautikpip** [comment]
Hi @mousedownmike,

As per my understanding tracing presign call should be a feature request. I think you are receiving this panic because you might have started xray tracing on S3 session and it tries to capture presign call (because presign call gets object from URL) which is not aws sdk call so xray won't be able to trace that call and generates panic. However, if you want to continue executing your program then you might want to set X Ray environment variable AWS_XRAY_CONTEXT_MISSING = LOG_ERROR. This is just a temporary work around to solve the panic issue.

**Comment by mousedownmike** [comment]
Hi @bhautikpip 

> I think you are receiving this panic because you might have started xray tracing on S3 session and it tries to capture presign call (because presign call gets object from URL) which is not aws sdk call so xray won't be able to trace that call and generates panic

I'm not sure I understand what you mean here.  Which part are you saying isn't an aws sdk call? 

My point was that if the same API works sometimes but fails at other times purely based on the data passed to it, it would seem like that's a bug.  Maybe it's just semantics and unimportant, I'm just wondering how best to communicate this so others don't run into the same problem.

From my sample repository, here is the call that will trigger the panic for large Body values but succeed for smaller values.

```go
output, err := c.Uploader().UploadWithContext(ctx, &s3manager.UploadInput{
		Body:   reader,
		Bucket: aws.String(c.Bucket),
		Key:    aws.String(to),
	})
```

Would it be safe to say that the [Upload Managers](https://docs.aws.amazon.com/sdk-for-go/api/service/s3/) described in the `sdk-for-go` documentation are not currently supported?  I'm happy to make a PR to the README to add some notes to that effect if I'm understanding the situation correctly.  Is `AWS_XRAY_CONTENT_MISSING` supported in `rc14`?

**Comment by mousedownmike** [comment]
Defaulting `AWS_XRAY_CONTENT_MISSING` to `LOG_ERROR` is the opposite of the default in the Java SDK.  I'm not sure how much consistency developers expect between SDKs. 

Here's the documentation on the Java side:

https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-java-configuration.html#xray-sdk-java-configuration-envvars

**Comment by shogo82148** [comment]
FYI, the default of `AWS_XRAY_CONTENT_MISSING` is `RUNTIME_ERROR` in NodeJS, Python, Ruby and .Net

https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-nodejs-configuration.html#xray-sdk-nodejs-configuration-envvars
https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-python-configuration.html#xray-sdk-python-configuration-envvars
https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-ruby-configuration.html#xray-sdk-ruby-configuration-envvars
https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-dotnet-configuration.html#xray-sdk-dotnet-configuration-envvars

**Comment by bhautikpip** [comment]
Hi @mousedownmike,

I was able to repro your issue. Can you provide your source URL of small files in which it's working ? looks like this handler (https://github.com/aws/aws-xray-sdk-go/blob/master/xray/aws.go#L60) receives empty unknown context and that's why it throws the panic you have mentioned. I am currently investigating on why that handler receives empty context. AWS_XRAY_CONTEXT_MISSING is not in the rc14 but will be in the latest release.

**Comment by mousedownmike** [comment]
@bhautikpip setting `SOURCE_URL` in that sample project I created to the following does not cause the panic:

https://blog.mikedalrymple.com/featured/gg-bridge-fog.jpg

**Comment by bhautikpip** [comment]
Hi @mousedownmike,

I think I know why small files are not receiving panic. So basically when file is large S3 APIs upload data in parts and when file is small S3 API uploads single file. So upload function for smaller files does not use presign request (https://github.com/aws/aws-sdk-go/blob/master/service/s3/s3manager/upload.go#L378) where as multi part upload (https://github.com/aws/aws-sdk-go/blob/6627e8938cdef5476dec0510b524258d3bf3983e/service/s3/s3manager/upload.go#L385)uses presign request (https://github.com/aws/aws-sdk-go/blob/6627e8938cdef5476dec0510b524258d3bf3983e/service/s3/s3manager/upload.go#L622) and presign request is causing problem I have mentioned in previous comment which leads to panic.

**Comment by mousedownmike** [comment]
That sounds right, and I can confirm that these operations succeed when I use the latest code in master (i.e. with the `LOG_ERROR`) set as the default.

**Comment by bhautikpip** [comment]
Yeah it is suppressing the panic and continue running your application.

**Comment by bhautikpip** [comment]
Hi @mousedownmike,

Based on my investigation Basically in this call (https://github.com/aws/aws-sdk-go/blob/6627e8938cdef5476dec0510b524258d3bf3983e/service/s3/s3manager/upload.go#L622) getReq call PresignRequest without passing context. since PresignRequest is a sign call it basically calls xray handlers and perform BeginSegment operation. Since it doesn't have a context passed on it generates this panic.

**Comment by mousedownmike** [comment]
I don't think this should be allowed to be closed until there's a fix for it.  It seems problematic to have x-ray code that will fail depending on the size of a payload being sent using an S3 SDK call.

**Comment by bhautikpip** [comment]
That's right. I will add this item in our backlog.

**Comment by swoldemi** [comment]
Just ran into this doing some late night experimenting:
CloudWatch logs:
```
START RequestId: 219e1b19-8e34-435b-b5d8-5f8c91367579 Version: $LATEST
2020-03-11T07:56:02Z [INFO] using daemon endpoints from environment variable AWS_XRAY_DAEMON_ADDRESS: 169.254.79.2:2000
2020-03-11T07:56:02Z [INFO] Emitter using address: 169.254.79.2:2000
2020/03/11 07:56:02 Attempting to download object swoldemi-tmp/test.img-7 with 2 CPU(s)
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 's3': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'attempt': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'unmarshal': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 's3': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'attempt': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 's3': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'attempt': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 's3': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'attempt': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 's3': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'attempt': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 's3': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'attempt': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 's3': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'attempt': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 's3': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'attempt': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 's3': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'attempt': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 's3': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'attempt': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'unmarshal': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'unmarshal': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'unmarshal': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'unmarshal': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'unmarshal': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'unmarshal': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'unmarshal': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'unmarshal': segment cannot be found.
2020-03-11T07:56:02Z [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'unmarshal': segment cannot be found.
2020/03/11 07:56:05 Downloaded test.img-7 bytes 293601280
2020/03/11 07:56:05 Completed downloading 293601280 bytes
END RequestId: 219e1b19-8e34-435b-b5d8-5f8c91367579
REPORT RequestId: 219e1b19-8e34-435b-b5d8-5f8c91367579	Duration: 3189.76 ms	Billed Duration: 3200 ms	Memory Size: 2048 MB	Max Memory Used: 1412 MB	Init Duration: 130.43 ms	
```
Lambda function in question:
https://gist.github.com/swoldemi/2b473b6746b8936d3f59069174cee9db

go.mod:
```
module github.com/swoldemi/downloadbenchmark

go 1.14

require (
	github.com/aws/aws-lambda-go v1.15.0
	github.com/aws/aws-sdk-go v1.29.21
	github.com/aws/aws-xray-sdk-go v1.0.0-rc.14
)
```

**Comment by bhautikpip** [comment]
Hi @swoldemi ,

Thanks for pointing this out. I think this is happening because aws-sdk-go code base creates multiple workers to download the object and calls aws-xray-sdk-go's handlers asynchronously. We are currently addressing the same issue with multi part upload. I will add this item in our backlog too.

**Comment by swoldemi** [comment]
Hello @bhautikpip. Thanks for the quick response. That makes sense. The number of errors are proportionate to the number of workers (unmarshal/attempt/s3 subsegments). Looking forward to the fix!
